File: src/main/java/engine/model/VertexArrayObject.java
package engine.model;

import org.lwjgl.opengl.*;
import renderer.OpenGLContext;

import java.lang.ref.WeakReference;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

public class VertexArrayObject {

    private final EnumSet<DataChannels> channels;
    private int id = -1;
    private transient boolean attributesSetUp = false;


    public static VertexArrayObject getForChannels(EnumSet<DataChannels> channels) {
        return new VertexArrayObject(channels);
    }

    private VertexArrayObject(EnumSet<DataChannels> channels) {
        this.channels = channels.clone();
        OpenGLContext.getInstance().execute(() -> setId(GL30.glGenVertexArrays()));
        setUpAttributes();
    }

    private void unbind() {
        OpenGLContext.getInstance().execute(() -> GL30.glBindVertexArray(0));
    }

    public void bind() {
//        if(CURRENTLY_BOUND_VAO == id) { return; }
        OpenGLContext.getInstance().execute(() -> {
            GL30.glBindVertexArray(getId());
        });
    }

    private void setUpAttributes() {
        if(attributesSetUp) { return; }
        OpenGLContext.getInstance().execute(() -> {
            bind();
            int currentOffset = 0;
            for (DataChannels channel : channels) {
                GL20.glEnableVertexAttribArray(channel.getLocation());
                GL20.glVertexAttribPointer(channel.getLocation(),channel.getSize(), GL11.GL_FLOAT, false, bytesPerVertex(channels), currentOffset);

                currentOffset += channel.getSize() * 4;
            }
        });
        attributesSetUp = true;
    }

    private static Map<EnumSet<DataChannels>, Integer> cache = new HashMap<>();

    public static int bytesPerVertex(EnumSet<DataChannels> channels) {
        if(cache.containsKey(channels)) {
            return cache.get(channels);
        } else {
            int sum = 0;
            for (DataChannels channel : channels) {
                sum += channel.getSize();
            }
            int bytesPerVertex = sum * 4;
            cache.put(channels, bytesPerVertex);
            return bytesPerVertex;
        }
    }

    public void delete() {
        GL30.glDeleteVertexArrays(id);
    }

    public int getBytesPerVertex() {
        return bytesPerVertex(channels);
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }
}

File: opengl/src/main/kotlin/de/hanno/hpengine/graphics/buffer/vertex/VertexArrayObject.kt
package de.hanno.hpengine.graphics.buffer.vertex

import de.hanno.hpengine.graphics.DataChannels
import de.hanno.hpengine.graphics.GraphicsApi
import org.lwjgl.opengl.*

import java.util.*


class VertexArrayObject(
    private val graphicsApi: GraphicsApi,
    private val channels: EnumSet<DataChannels>
) {

    var id = graphicsApi.onGpu { GL30.glGenVertexArrays() }
    init {
        setUpAttributes()
    }

    private fun setUpAttributes() = graphicsApi.onGpu {
        bind()
        var currentOffset = 0
        for (channel in channels) {
            GL20.glEnableVertexAttribArray(channel.location)
            GL20.glVertexAttribPointer(channel.location, channel.size, GL11.GL_FLOAT, false, bytesPerVertex(channels), currentOffset.toLong())

            currentOffset += channel.size * 4
        }
    }

    fun bind() = graphicsApi.onGpu {
        GL30.glBindVertexArray(id)
    }

    fun delete() = GL30.glDeleteVertexArrays(id)

    companion object {
        fun getForChannels(graphicsApi: GraphicsApi, channels: EnumSet<DataChannels>) = VertexArrayObject(graphicsApi, channels)

        private val cache = HashMap<EnumSet<DataChannels>, Int>()

        fun bytesPerVertex(channels: EnumSet<DataChannels>): Int = if (cache.containsKey(channels)) {
            cache[channels]!!
        } else {
            var sum = 0
            for (channel in channels) {
                sum += channel.size
            }
            val bytesPerVertex = sum * 4
            cache[channels] = bytesPerVertex
            bytesPerVertex
        }
    }
}
